cmd/argoexec/commands/artifact/delete.go

```diff
@@ -52,63 +53,121 @@ func NewArtifactDeleteCommand() *cobra.Command {
 	}
 }
 
-func deleteArtifacts(labelSelector string, ctx context.Context, artifactGCTaskInterface wfv1alpha1.WorkflowArtifactGCTaskInterface) error {
+type request struct {
+	Task             *v1alpha1.WorkflowArtifactGCTask
+	NodeName         string
+	ArtifactNodeSpec *v1alpha1.ArtifactNodeSpec
+}
+
+type response struct {
+	Task     *v1alpha1.WorkflowArtifactGCTask
+	NodeName string
+	Results  map[string]v1alpha1.ArtifactResult
+	Err      error
+}
 
+func deleteArtifacts(labelSelector string, ctx context.Context, artifactGCTaskInterface wfv1alpha1.WorkflowArtifactGCTaskInterface) error {
 	taskList, err := artifactGCTaskInterface.List(context.Background(), metav1.ListOptions{LabelSelector: labelSelector})
 	if err != nil {
 		return err
 	}
+	taskWorkers := env.LookupEnvIntOr(common.EnvExecGCWorkers, 4)
+
+	totalTasks := 0
+	for _, task := range taskList.Items {
+		totalTasks += len(task.Spec.ArtifactsByNode)
+	}
+	taskQueue := make(chan *request, totalTasks)
+	responseQueue := make(chan response, totalTasks)
+	for i := 0; i < taskWorkers; i++ {
+		go deleteWorker(ctx, taskQueue, responseQueue)
+	}
 
+	taskno := 0
+	nodesToGo := make(map[*v1alpha1.WorkflowArtifactGCTask]int)
 	for _, task := range taskList.Items {
+		nodesToGo[&task] = len(task.Spec.ArtifactsByNode)
 		task.Status.ArtifactResultsByNode = make(map[string]v1alpha1.ArtifactResultNodeStatus)
 		for nodeName, artifactNodeSpec := range task.Spec.ArtifactsByNode {
-
-			var archiveLocation *v1alpha1.ArtifactLocation
-			artResultNodeStatus := v1alpha1.ArtifactResultNodeStatus{ArtifactResults: make(map[string]v1alpha1.ArtifactResult)}
-			if artifactNodeSpec.ArchiveLocation != nil {
-				archiveLocation = artifactNodeSpec.ArchiveLocation
+			taskQueue <- &request{Task: &task, NodeName: nodeName, ArtifactNodeSpec: &artifactNodeSpec}
+			taskno++
+		}
+	}
+	close(taskQueue)
+	completed := 0
+	for {
+		response := <-responseQueue
+		if response.Err != nil {
+			return response.Err
+		}
+		if response.Task == nil {
+			completed++
+			if completed >= taskWorkers {
+				break
 			}
+		} else {
+			response.Task.Status.ArtifactResultsByNode[response.NodeName] = v1alpha1.ArtifactResultNodeStatus{ArtifactResults: response.Results}
+			// Check for completed tasks
+			nodesToGo[response.Task]--
```

Comment:
Do you think we need to add an ok check here? I see that this should obviously exist, just tempted to be more defensive so that we future proof this a bit more. 

Alternatively add a comment that we expect the response.Task to be present here because it was populated in the previous loop, probably should be accompanied by a comment at line 89 where it gets populated. 

Sorry for being pedantic/annoying when it is obvious, I just don't want things to go in the way of NodeStatus, tempted to use a linter to force ok checks everywhere to be honest.

Reply:
Thanks!

Reply:
sorry, is this the `if response.Task == nil` check? I'm actually kind of confused why that's in there. It seems redundant, no?